### ISSUE TITLE
Panic when initializing tracer with prometheus default registerer: `duplicate metrics collector registration attempted`

### ISSUE THREAD

**Issue by masroorhasan** [jaegertracing/jaeger-lib#48]
## Requirement - what kind of business use case are you trying to solve?

Attempting to enable client side prometheus metrics which blows up application code (See below). We need metrics to track dropped spans, etc. 

Jaeger Go Client: v2.14.0
Jaeger Lib: v1.5.0

## Problem - what in Jaeger blocks you from solving the requirement?
**tl;dr** Initializing new tracer with prometheus `DefaultRegisterer` is panicking with `panic: duplicate metrics collector registration attempted`.

Upon initializing tracer with the default prometheus registerer, by doing either:
```
factory := jprom.New(jprom.WithRegisterer(prometheus.DefaultRegisterer))
return cfg.NewTracer(
    jaegercfg.Metrics(factory),
)
```

or (from example - both should be exactly the same):
```
factory := jprom.New()
return cfg.NewTracer(
	jaegercfg.Metrics(factory),
)
```

I'm running into the following panic from the `MustRegister` [call]( https://github.com/jaegertracing/jaeger-lib/blob/771be3fa752608465734addac8d70763c5601eda/metrics/prometheus/cache.go#L50) from the metrics cache code which is used to [initialize](https://github.com/jaegertracing/jaeger-lib/blob/fe748a21c8698beaae2a738e4e1d0b6ba53b254f/metrics/metrics.go#L72) all metrics (counters, gauges, etc) via reflection from metrics defined [here](https://github.com/jaegertracing/jaeger-client-go/blob/master/metrics.go). 

The error and conditions of the error is described in prometheus [docs](https://godoc.org/github.com/prometheus/client_golang/prometheus#AlreadyRegisteredError).
```
panic: duplicate metrics collector registration attempted

goroutine 1 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc420083500, 0xc4203f9790, 0x1, 0x1)
	/go/src/github.com/prometheus/client_golang/prometheus/registry.go:404 +0x9e
github.com/uber/jaeger-lib/metrics/prometheus.(*vectorCache).getOrMakeCounterVec(0xc420366300, 0x0, 0x0, 0x0, 0x0, 0xc4203e6640, 0xd, 0xc4203e6640, 0xd, 0x0, ...)
	/go/src/github.com/uber/jaeger-lib/metrics/prometheus/cache.go:50 +0x213
github.com/uber/jaeger-lib/metrics/prometheus.(*Factory).Counter(0xc4203dc780, 0xc4203e6640, 0xd, 0xc4203663f0, 0xc42028b298, 0x2)
	/go/src/github.com/uber/jaeger-lib/metrics/prometheus/factory.go:136 +0x186
github.com/uber/jaeger-lib/metrics.initMetrics(0xb098a0, 0xc420242c00, 0xd86f00, 0xc4203dc780, 0x0, 0x6, 0x6)
	/go/src/github.com/uber/jaeger-lib/metrics/metrics.go:72 +0x43e
github.com/uber/jaeger-lib/metrics.Init(0xb098a0, 0xc420242c00, 0xd86f00, 0xc4203dc780, 0x0)
	/go/src/github.com/uber/jaeger-lib/metrics/metrics.go:25 +0x57
github.com/uber/jaeger-client-go.NewMetrics(0xd86f00, 0xc4203dc730, 0x0, 0xd86f00)
	/go/src/github.com/uber/jaeger-client-go/metrics.go:100 +0xa5
github.com/uber/jaeger-client-go/config.Configuration.NewTracer(0xc42003a074, 0xa, 0x0, 0x0, 0x0, 0x0, 0xc42025bc40, 0xc420366210, 0x0, 0x0, ...)
	/go/src/github.com/uber/jaeger-client-go/config/config.go:178 +0xef

... application stack
```

Now our application code defines a ton of metrics that are already registered to the `DefaultRegisterer` at this point but there is no way to tell if an already registered metrics' `Desc` is clashing with jaeger client [metrics](https://github.com/jaegertracing/jaeger-client-go/blob/master/metrics.go) or there is a bug in this reflection code which is re-registering one of the metrics twice (there is some hashing going on to create the cache key).

## Proposal - what do you suggest to solve the problem or improve the existing situation?
**tl;dr**: Use `prometheus.Register` instead of `prometheus.MustRegister` so errors can be logged or applied to use in business logic.

Couple of proposals:

~~1. Add test cases either here or in jaegertracing/jaeger-client-go to test for duplicate metrics registration (apologies if this is already done - i havent checked if these tests exist).~~

2. Change the signature of the functions in [cache.go](https://github.com/jaegertracing/jaeger-lib/blob/771be3fa752608465734addac8d70763c5601eda/metrics/prometheus/cache.go) to return error object on failure:
```
func (c *vectorCache) getOrMakeCounterVec(opts prometheus.CounterOpts, labelNames []string) (*prometheus.CounterVec, error) {
// ...
``` 
This can be done by switching from using `c.registerer.MustRegister(cv)` ([docs](https://godoc.org/github.com/prometheus/client_golang/prometheus#Registry.MustRegister)) to `err := c.registerer.Register(cv)` ([docs](https://godoc.org/github.com/prometheus/client_golang/prometheus#Register)) and return the error to calling code. The error should also be logged to give more verbosity to application user. Obviously, repeated changes for `getOrMakeGaugeVec` and `getOrMakeHistogramVec` as well. 

The reason for this second proposal is because its impossible to find the offending metric definition in hundreds/thousands of metrics defined both in user application and library code (jaeger-client-go). Logging the offending metric would at least let the user more information. 

## Any open questions to address

Should we skip registering an offending metric if registering the said metric with `getOrMake*Vec` (Counter, Gauge, Timer) returns an error?

**Comment by yurishkuro** [comment]
> duplicate metrics **collector** registration

This looks suspicious. I don't think it has to do with duplicate metric names (although it might). If I run the HotROD example 
```
$ go run ./examples/hotrod/main.go all --metrics prometheus
```

I am getting all kinds of RPC and Jaeger metrics without conflicts (http://localhost:8083/metrics).

**Comment by yurishkuro** [comment]
I would recommend trying out your approach by changing the code in vendor directly. I haven't looked too closely at the code, but somehow I think that if Register() would return an error, then MustRegister() would panic with that exact same error, which in this case does not tell which metrics is clashing. However, if we get the error in back in Jaeger code which knows that it was _trying_ to register, then we can panic with a more description error.

**Comment by masroorhasan** [comment]
@yurishkuro Good idea - just made the following change on `getOrMakeCounterVec`:
```
err := c.registerer.Register(cv)
                if err != nil {
                        panic(fmt.Sprintf("error registering metric with name:'%s', with labels '%v', error: '%v'", opts.Name, labelNames, err))
                }
```

This is the panic output:
`
panic: error registering metric with name:'jaeger_traces', with labels '[sampled state]', error: 'duplicate metrics collector registration attempted'
`

As shown in the panic, the metric its complaining about is `jaeger_traces`, this kind of points to the metrics defined in the jaeger client - I believe the logic prepends `jaeger_` to each metric name (not 100% on that but i can look around). Here are the metrics named (or metadata) with `traces` defined in jaeger client:
```
// Number of traces started by this tracer as sampled
	TracesStartedSampled metrics.Counter `metric:"traces" tags:"state=started,sampled=y"`

	// Number of traces started by this tracer as not sampled
	TracesStartedNotSampled metrics.Counter `metric:"traces" tags:"state=started,sampled=n"`

	// Number of externally started sampled traces this tracer joined
	TracesJoinedSampled metrics.Counter `metric:"traces" tags:"state=joined,sampled=y"`

	// Number of externally started not-sampled traces this tracer joined
	TracesJoinedNotSampled metrics.Counter `metric:"traces" tags:"state=joined,sampled=n"`
```

**Comment by yurishkuro** [comment]
yes, but they have different label values, and uniqueness in Prometheus is determined on the name + labels.

**Comment by masroorhasan** [comment]
@yurishkuro makes sense, it was because of how we initialized the jaeger tracer twice from our grpc and gateway servers (both using the same prometheus registry and pointing to same metrics endpoint). 

Thanks for your responses, closing the issue as its not a problem on jaeger-lib.
